diff --git a/drivers/gpu/drm/i915/i915_drv.c b/drivers/gpu/drm/i915/i915_drv.c
index 69d8ed5..40927a0 100644
--- a/drivers/gpu/drm/i915/i915_drv.c
+++ b/drivers/gpu/drm/i915/i915_drv.c
@@ -428,6 +428,11 @@ void intel_detect_pch(struct drm_device *dev)
 				DRM_DEBUG_KMS("Found LynxPoint LP PCH\n");
 				WARN_ON(!IS_HASWELL(dev));
 				WARN_ON(!IS_ULT(dev));
+            } else if (id == 0x2918) {
+                /* INTEL ICH9 LPC - Apply CougarPoint settings */
+                dev_priv->pch_type = PCH_CPT;
+                DRM_DEBUG_KMS("Found CougarPoint PCH\n");
+                WARN_ON(!(IS_GEN6(dev) || IS_IVYBRIDGE(dev)));
 			} else {
 				goto check_next;
 			}
